src/digest/digest.rs

```diff
@@ -283,8 +283,13 @@ pub struct Algorithm {
     /// The length of the length in the padding.
     len_len: usize,
 
+    #[cfg(not(feature="no_asm"))]
     block_data_order: unsafe extern fn(state: &mut [u64; MAX_CHAINING_LEN / 8], data: *const u8,
                                        num: c::size_t),
+    #[cfg(feature="no_asm")]
+    block_data_order: unsafe fn(state: &mut [u64; MAX_CHAINING_LEN / 8], data: *const u8,
```

Comment:
In the no_asm case, we could actually make `block_data_order` a normal, non-`unsafe` function, right? In the SHA-1 code, we could skip over the unsafe -> safe adapter function, for example.